base/system/winlogon/sound.c

```diff
@@ -0,0 +1,317 @@
+
+    if (bLogon)
+    {
+        SERVICE_STATUS_PROCESS Info;
+        DWORD dwBytesNeeded;
+        ULONG Index = 0;
+        SC_HANDLE hSCManager, hService;
+
+        /* Open the service manager */
+        hSCManager = OpenSCManagerW(NULL, NULL, SC_MANAGER_CONNECT);
+        if (!hSCManager)
+        {
+            ERR("OpenSCManagerW failed (%x)\n", GetLastError());
+            goto Cleanup;
+        }
+
+        /* Open the wdmaud service */
+        hService = OpenServiceW(hSCManager, L"wdmaud", GENERIC_READ);
+        if (!hService)
+        {
+            /* The service is not installed */
+            TRACE("Failed to open wdmaud service (%x)\n", GetLastError());
+            CloseServiceHandle(hSCManager);
+            goto Cleanup;
+        }
+
+        /* Wait for wdmaud to start */
+        do
+        {
+            if (!QueryServiceStatusEx(hService, SC_STATUS_PROCESS_INFO, (LPBYTE)&Info, sizeof(Info), &dwBytesNeeded))
+            {
+                TRACE("QueryServiceStatusEx failed (%x)\n", GetLastError());
+                break;
+            }
+
+            if (Info.dwCurrentState == SERVICE_RUNNING)
+            {
+                break;
+            }
+
+            Sleep(1000);
+
+        } while (Index++ < 20);
+
+        CloseServiceHandle(hService);
+        CloseServiceHandle(hSCManager);
+
+        /* If wdmaud is not running exit */
+        if (Info.dwCurrentState != SERVICE_RUNNING)
+        {
+            WARN("wdmaud has not started!\n");
+            goto Cleanup;
+        }
+
+        /* Sound subsystem is running. Play logon sound. */
+    }
+
+    /* Impersonate current user */
+    if (!ImpersonateLoggedOnUser(PSData->Session->UserToken))
+    {
+        goto Cleanup;
+    }
+
+    /* Open user's HKCU */
+    lError = RegOpenCurrentUser(KEY_READ, &hUserKey);
+
+    RevertToSelf();
+
+    if (lError != ERROR_SUCCESS)
+    {
+        goto Cleanup;
+    }
+
+    switch (PSData->Sound)
+    {
+        case SYSTEMSND_LOGON:
+            pszSndKey = L"AppEvents\\Schemes\\Apps\\.Default\\WindowsLogon\\.Current";
+            break;
+
+        case SYSTEMSND_LOGOFF:
+            pszSndKey = L"AppEvents\\Schemes\\Apps\\.Default\\WindowsLogoff\\.Current";
+            break;
+
+        case SYSTEMSND_ASTERISK:
+            pszSndKey = L"AppEvents\\Schemes\\Apps\\.Default\\SystemAsterisk\\.Current";
+            break;
+
+        case SYSTEMSND_EXCLAMATION:
+            pszSndKey = L"AppEvents\\Schemes\\Apps\\.Default\\SystemExclamation\\.Current";
+            break;
+
+        case SYSTEMSND_CRITICAL_STOP:
+            pszSndKey = L"AppEvents\\Schemes\\Apps\\.Default\\SystemHand\\.Current";
+            break;
+
+        case SYSTEMSND_QUESTION:
+            pszSndKey = L"AppEvents\\Schemes\\Apps\\.Default\\SystemQuestion\\.Current";
+            break;
+
+        default:
+            pszSndKey = L"AppEvents\\Schemes\\Apps\\.Default\\SystemDefault\\.Current";
+    }
```

Comment:
It looks like all that could be condensed (at the expense of re-introducing a sized string buffer as in the original code) into one single:
```c
    WCHAR szSndKey[MAX_PATH] = {0};

...

    StringCchCopyW(szSndKey, _countof(szSndKey),
                   L"AppEvents\\Schemes\\Apps\\.Default\\%s\\.Current",
                   SoundName);
```
with `SoundName` being one of these e.g. `"SystemQuestion"` etc. (see original code), to be passed in parameter, or, obtained from a mapping table:
```c
static LPCWSTR SndToEvtName[] = {
    L"SystemDefault",
    L"WindowsLogon",
    L"WindowsLogoff",
    L"SystemAsterisk",
    L"SystemExclamation",
    L"SystemHand",
    L"SystemQuestion"
};
```
that is then used as `SndToEvtName[PSData->Sound]` instead of the `SoundName` parameter in the `StringCchCopyW` call in the example above.

Reply:
With this implementation you don't need to copy any string and I think it's not a good idea to re-introduce the sized string buffer.

Reply:
The idea of using a mapping table sounds better.

Reply:
Normally it had to be the job of PlaySoundW to do the whole "AppEvents\\Schemes\\Apps\\..." rigamarole (and correctly expanding the paths etc.), but here we are working around our current limited implementation of it (that I discussed iirc. with Oleg a long time ago) and implement what it should do here instead in winlogon....
so whatever...

Reply:
Replace `HKEY_CURRENT_USER` (predefined key) in `PlaySoundW` implementation with `RegOpenCurrentUser` should work but still not working if you use `SND_ASYNC` flag because of this behavior.

> Prior to Windows XP SP2 and Windows Server 2003,
the new thread would be assigned the DACL from the impersonation token of the creator if the calling thread is impersonating or from the primary token if it is not. **Starting with Windows XP SP2 and Windows Server 2003 the behavior is always the same: the new thread is assigned the DACL from the primary token regardless of the impersonation state of the caller**.

[https://comp.os.ms-windows.programmer.win32.narkive.com/rgtREkqp/creating-new-threads-from-an-impersonated-thread](https://comp.os.ms-windows.programmer.win32.narkive.com/rgtREkqp/creating-new-threads-from-an-impersonated-thread)

But is it worth fixing that? `PlaySoundW` function is rarely called in system processes anyway.